minor touch-ups (#4241)
dotnet/csharplang: proposals/discriminated-unions.md

# Discriminated unions / `enum class`

`enum class`es are a new kind of type declaration, sometimes referred to as discriminated unions,

> where every possible instance of the type is listed, and instances do not overlap.

Reviewer: ```suggestion
where every possible instance of the type is listed, and each instance is non-overlapping.
```

Author: how about 'no two instances overlap'?

Reviewer: I prefer the original wording.

This is almost identical to the semantics of a top-level `enum class`, except that
the nested enum class defines a nested root type, and everything below the nested enum

> class is a subtype of the nested root type, rather than of the top-level root type.

Reviewer: Please revert this bit.

### Switch expression and patterns

There are some proposed adjustments to pattern matching and the `switch` expression to handle
`enum classes`. Switch expressions can already match types through the variable pattern, but
for reference types, currently no set of switch arms in the switch expression are considered
complete, except for matching against the static type of the argument, or a subtype.

> Switch expressions would be changed such that if the root type of an `enum class` is the static

Reviewer: Please revert this one.

Author: Ask some other writers what they think?

type of the argument to the switch expression, and there is a set of patterns matching all
members of the enum, then the switch will be considered exhaustive.

Since value members are not constants and do not define new static types, they currently cannot
be matched by pattern. To make this possible, a new pattern using the constant pattern syntax

> will be added to allow matching against `enum class` value members. A match is defined to succeed

Reviewer: ```suggestion
will be added to allow matching against `enum class` value members. The match is defined to succeed
```

if and only if the argument to the pattern match and the value returned by the `enum class` value

> member are reference equal, although the implementation is not required to perform this check.

Reviewer: Please revert this bit.

Author: I think it's better if I just withdraw this, and you do what you want.

Reviewer: If you want to. There are still some fixes left we'd be happy to take.
